### PR TITLE
ci: Add a portable testing framework and scalability tests [SCALE-29]

### PR DESCRIPTION
**Pull request by highvelcty**
## Description
Add the Determined AI Scalability Testing (daist) to the determined repository. This is python `unittest` based testing against `determined` deployments. See the README.md within these changes.

Most of the development of `daist` took place in the determined-ai/scalability repo. The code was moved from [here](https://github.com/determined-ai/scalability/tree/a4d46c2e2739ae580997a9eaa9b595bac4c913bc/tests). 

## Test Plan
The code within this PR is not packaged in `determined`, it provides testing for `determined`. No further testing is needed for a `determined` release in regard to these changes.

The `daist.rest_api.TestRO` test is executed as part of the circleci test-e2e/test-perf job. This job is executed on the `main` branch only by default.

Part of this effort is to port "K6" rest API testing to [`Locust`](https://locust.io/) based testing with `daist`.. Regression testing between K6/daist at various commits was made:

| Commit | Original Date | K6 Avg Latency (ms) | Locust Avg Latency (ms) | Delta (ms) |
| -----  | ------------- | -------------- | ----------------------- |--------------- |
| 6cd7d060b0fb63a6557eecccf05c812718e5fe38 | 2024-06-17 15:20:27.301489+00 | 664 | 735 | 71    |
| f721751fa5851b93c2108ddc6b0b8a6c4909ff77 | 2024-07-22 20:29:49.138596+00 | 772 | 730 | -42   |
| 0f8e90e55f3b68ed92dd94152423815e70e97dea | 2024-07-29 18:22:50.82178+00 | 687 | 772 | 85    |

Additionally, determined commit  95b5164190e8378f212ed8ed5ff04beec7ba0ae6 had a problem that was previously caught with K6 testing. This commit was tested and the problem was also caught with `daist` testing.

image

## Checklist

- [x] Changes have been manually QA'd
- [x] New features have been approved by the corresponding PM
- [N/A] User-facing API changes have the "User-facing API Change" label
- [N/A] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master
/docs/release-notes/README.md) for details.
  - These changes are not packaged with determined, therefore no release notes were made.
- [N/A] Licenses have been included for new code which was copied and/or modified from any external code

Internal change types:
- test: new tests

Coauthor: Sean Mackrory

[SCALE-29]: https://hpe-aiatscale.atlassian.net/browse/SCALE-29?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .circleci/real_config.yml [hunk 1]
```diff
@@ -282,7 +282,7 @@ commands:
       - when:
           condition: <<parameters.tf2>>
           steps:
-            - run: docker pull determinedai/pytorch-ngc-dev:5432424
+            - run: docker pull determinedai/pytorch-ngc-dev:f20b027
```
[the remark this answers is not in the document]
Comment (highvelcty) [reply .circleci/real_config.yml]: Good catch. No, I did not intend to change the image hash. Fixed.

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -10,7 +10,7 @@
 /go.sum         @determined-ai/backend
 /agent          @determined-ai/backend
 /proto          @determined-ai/backend
-/performance    @determined-ai/web
+/performance    @determined-ai/scalability
```
[the remark this answers is not in the document]
Comment (highvelcty) [reply .github/CODEOWNERS:13]: @mackrorysd , will you please comment here.

File: performance/daist/uts/framework/test_log.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import Tuple
+from unittest import TestCase
+
+from daist.framework import log
+from daist.framework.paths import PkgPath
+
+
+class Test(TestCase):
```
[the remark this answers is not in the document]
Comment (highvelcty) [reply performance/daist/uts/framework/test_log.py:13]: OK, just some background; this is almost nothing more than applying the python `unittest` framework that is part of the standard library. It has a  been extended to provide consistent logging across all tests and structured output from the tests.

The `unittest` framework is being used to provide "junit-like' addressing to the testing execution entry points. e.g. `daist.metrics.test_latency`  and `daist.rest_api.test_locust`. Furthermore, it provides a means to group tests into suites as necessary.

File: performance/daist/uts/models/samples/PerfTestRun-LocustTest.test_concurrent_read_only_users.txt [hunk 4]
```diff
@@ -0,0 +1,59 @@
+PerfTestRunsTable:
```
[the remark this answers is not in the document]
Comment (highvelcty) [reply performance/daist/uts/models/samples/PerfTestRun-LocustTest.test_concurrent_read_only_users.txt:1]: yes, this is a sample for use in unit testing the `daist` package only. It is the only binary file, it is small (12599 bytes) and unchanging.

File: .github/CODEOWNERS (hunk 2 above)
[the remark this answers is not in the document]
Comment (mackrorysd) [reply .github/CODEOWNERS:13]: I'm going to keep looking at the data regularly, so I was gonna keep it on my radar for now. Will need to revisit in a few months and see what long-term staffing looks like.

**Review by djanicekpach — APPROVED** [review]

**Review by hkang1 — APPROVED** [review]
I took a look as web's POV and I don't think there are any objections to backend (or scalability team) owning locust API tests. I think it's more important that web gets notified when the swagger generated api-ts-sdk changes for determined (which is does not intersect with this PR at all). The changes here look like it's entirely backend API testing with locust with scaling as a focus, so LGTM!

**Comment by highvelcty** [comment]
> I think it's more important that web gets notified when the swagger generated api-ts-sdk changes for determined (which is does not intersect with this PR at all).

It is good to know that testing the typescript SDK is desirable. 

Just a note, neither the legacy K6 testing nor the new locust testing test the typescript SDK. I mention this to be clear that there is not loss of testing coverage with the transition.

File: performance/daist/daist/rest_api/locust_utils.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+    HTTPConnection.debuglevel = 1
+    logging.basicConfig()
+    logging.getLogger().setLevel(logging.DEBUG)
+    requests_log = logging.getLogger("requests.packages.urllib3")
+    requests_log.setLevel(logging.DEBUG)
+    requests_log.propagate = True
+
+
+class LocustTasksWithMeta(UserList):
```
[the remark this answers is not in the document]
Comment (highvelcty) [reply performance/daist/daist/rest_api/locust_utils.py:29]: Right on, thanks for the heads up. I made a comment in the code about this for future work. At this time, I would like to leave it functionally as it is for this merge.

**Pull request merged**